### PR TITLE
[DEV-4478] Migrating away from Custom Client Side Router to React Router w/o the Hash

### PR DESCRIPTION
**Pull request by MaxwellKendall**
**High level description:**
Implements a few high level changes; most notably refactored (a) the Glossary and (b) Search Containers along with our (c) general routing logic and sitemap generation.

**Technical details:**
- Glossary & Search Container were the main thing to be updated as they dealt with the url quite a bit
- Everywhere we programmatically navigate had to be updated 

**JIRA Ticket:**
[DEV-4478](https://federal-spending-transparency.atlassian.net/browse/DEV-4478)

The following are ALL required for the PR to be merged:

Author:
- [x] Linked to this PR in JIRA ticket
- [x] Adequately Tested in Sandbox
- [x] OPS Configuration for DEV ENV to return Custom Error Response for CloudFront 404s
- [x] Scheduled demo including Design/Testing/Front-end OR Provided instructions for testing in JIRA and PR `if applicable`
- [x] Verified cross-browser compatibility: Chrome, Safari, Firefox, Internet Explorer, Edge
`N/A` Verified mobile/tablet/desktop/monitor responsiveness
- [x] Verified that this PR does not create any *new* accessibility issues (via Axe Chrome extension)
- [x] Added Unit Tests for methods in Container Components, reducers, helper functions, and models `if applicable`
`N/A` All `componentWillReceiveProps` and `componentWillMount` in relevant child/parent components/containers replaced with [future compatible life-cycle methods](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html) per [JIRA item 3277](https://federal-spending-transparency.atlassian.net/browse/DEV-3277)
`N/A` [API contract](https://github.com/fedspendingtransparency/usaspending-api/tree/dev/usaspending_api/api_contracts) updated `if applicable`
`N/A` [Component Library Integration Status Report](https://github.com/fedspendingtransparency/data-act-documentation/blob/data-transparency-ui/frontend_apps/component-library-integration-status.md) updated `if applicable`

Reviewer(s):
`N/A` Design review complete `if applicable`
`N/A` [API #1234](https://github.com/fedspendingtransparency/usaspending-api/pull/1234) merged concurrently `if applicable`
- [x] Code review complete

### CODE REVIEW

File: src/js/components/sharedComponents/Loading.jsx [hunk 1]
```diff
@@ -39,10 +45,25 @@ export const LoadingWrapper = ({
     if (isLoading) {
         return (
             // Not spending too much time w/ styles here, will have scope to improve later.
```
Comment (jonhill13) [inline src/js/components/sharedComponents/Loading.jsx:47]: Can we remove this comment or make it a `TODO`?

File: src/js/components/sharedComponents/header/NavBar.jsx [hunk 2]
```diff
@@ -160,13 +161,16 @@ export default class NavBar extends React.Component {
                             <li
                                 className="full-menu__item"
                                 role="menuitem">
-                                <a
+                                <Link className="full-menu__link" to="/explorer" onClick={clickedHeaderLink.bind(null, '/explorer')}>
+                                    Spending Explorer
+                                </Link>
+                                {/* <a
```
Comment (jonhill13) [inline src/js/components/sharedComponents/header/NavBar.jsx:167]: Do we still want this commented code?

File: src/js/components/sharedComponents/header/DropdownItem.jsx [hunk 3]
```diff
@@ -59,14 +60,14 @@ export default class DropdownItem extends React.Component {
         }
 
         let link = (
-            <a
+            <Link
                 className={`nav-children__link ${className}`}
-                href={this.props.url}
+                to={this.props.url}
                 onClick={clickedHeaderLink.bind(null, `${this.props.url.replace('#', '')}`)}
```
Comment (jonhill13) [inline src/js/components/sharedComponents/header/DropdownItem.jsx:66]: Do we still need this replace?

File: src/js/containers/search/SearchSidebarSubmitContainer.jsx [hunk 4]
```diff
@@ -7,13 +7,13 @@ import React from 'react';
...
 import { bindActionCreators } from 'redux';
 import { connect } from 'react-redux';
-import { is } from 'immutable';
 
 import * as appliedFilterActions from 'redux/actions/search/appliedFilterActions';
 import { clearAllFilters as clearStagedFilters } from 'redux/actions/search/searchFilterActions';
 import { setCheckedNaics, setUncheckedNaics } from 'redux/actions/search/naicsActions';
 import { resetMapLegendToggle } from 'redux/actions/search/mapLegendToggleActions';
 
+import { areFiltersEqual } from 'containers/search/SearchContainer';
```
Comment (jonhill13) [inline src/js/containers/search/SearchSidebarSubmitContainer.jsx:16]: Can we move this to a helper function?

File: src/js/components/explorer/landing/ExplorerLanding.jsx [hunk 5]
```diff
@@ -35,10 +36,10 @@ export default class ExplorerLanding extends React.Component {
...
...        The data powering the Spending Explorer is reported through a newly implemented law (the Digital Accountability and Transparency Act of 2014 (<a href="https://www.gpo.gov/fdsys/pkg/PLAW-113publ101/html/PLAW-113publ101.htm" target="_blank" rel="noopener noreferrer">DATA Act</a>)) that requires agencies to link their financial data, in addition to the award data that agencies were already reporting under the Federal Funding Accountability and Transparency Act of 2006 (FFATA). This data was first collected in the second quarter of fiscal year 2017. The data displayed on the site may not be complete or match official publications (e.g., the President&apos;s Budget, Agency Financial Reports), due to agencies&apos; different reporting schedules and the fact that some data elements aren&apos;t required to be reported. Notably, the Department of Defense (DoD) doesn&apos;t report their financial data using the same schedule as other major agencies; the DATA Act gives them an exemption that allows them to report later.  That said, we publish all the data agencies submit to provide as comprehensive a view as possible, and we expect available data to become more complete in the future.
```
[the remark this answers is not in the document]
Comment (MaxwellKendall) [reply src/js/components/explorer/landing/ExplorerLanding.jsx:45]: IF its an external link we use an a tag.

File: jest.config.js [hunk 6]
```diff
@@ -1,3 +1,6 @@
+// TODO: [DEV-5897] USAS Front End Test Configuration Update: Functional Components & Redux
```
[the remark this answers is not in the document]
Comment (MaxwellKendall) [reply jest.config.js:1]: I think it's nice to keep it.

File: src/js/components/agency/overview/AgencyOverview.jsx [hunk 7]
```diff
@@ -219,7 +219,7 @@ export default class AgencyOverview extends React.PureComponent {
                         }}>
                         <h4>
                             Budgetary Resources
-                            <a href={`#/agency/${this.props.agency.id}?glossary=budgetary-resources`}>
+                            <a href={`agency/${this.props.agency.id}?glossary=budgetary-resources`}>
```
[the remark this answers is not in the document]
Comment (MaxwellKendall) [reply src/js/components/agency/overview/AgencyOverview.jsx:222]: Sure

File: src/js/components/covid19/DataSourcesAndMethodologiesPage.jsx [hunk 8]
```diff
@@ -190,7 +192,7 @@ export default () => {
...
+                                            <strong>Details:</strong> The Broker contains financial data sourced from agency financial systems. This information is submitted directly by federal agencies to the Broker on a monthly basis and is packaged with related award and subaward data by the Broker and certified by a Senior Accountable Official. Directly submitted financial data files cover: 1) Treasury Account Symbol (TAS) balances (in Broker File A); 2) breakdowns of TAS-level spending by several dimensions (in Broker File B); and 3) a further breakdown of the award component of that spending by each federal award (in Broker File C). The latter two of these files are used on this page; visit our <a href="/download_center/custom_account_data">Custom Account Data</a> page to download them: &quot;Account Breakdown by Program Activity &amp; Object Class&quot; (for File B) and &quot;Account Breakdown by Award&quot; (for File C).
```
[the remark this answers is not in the document]
Comment (MaxwellKendall) [reply src/js/components/covid19/DataSourcesAndMethodologiesPage.jsx:195]: Nice.

**Pull request merged**